Lib/multiprocessing/spawn.py

```diff
@@ -100,34 +100,31 @@ def spawn_main(pipe_handle, parent_pid=None, tracker_fd=None):
 
         if parent_pid is not None:
             source_process = _winapi.OpenProcess(
-                _winapi.PROCESS_DUP_HANDLE, False, parent_pid)
+                _winapi.PROCESS_ALL_ACCESS, False, parent_pid)
         else:
             source_process = None
-        try:
-            new_handle = reduction.duplicate(pipe_handle,
-                                             source_process=source_process)
-        finally:
-            if source_process is not None:
-                _winapi.CloseHandle(source_process)
+        new_handle = reduction.duplicate(pipe_handle,
+                                         source_process=source_process)
         fd = msvcrt.open_osfhandle(new_handle, os.O_RDONLY)
+        parent_sentinel = source_process
     else:
         from . import resource_tracker
         resource_tracker._resource_tracker._fd = tracker_fd
-        fd = pipe_handle
-    exitcode = _main(fd)
+        fd = parent_sentinel = pipe_handle
+    exitcode = _main(fd, parent_sentinel)
     sys.exit(exitcode)
 
 
-def _main(fd):
-    with os.fdopen(fd, 'rb', closefd=True) as from_parent:
+def _main(fd, parent_sentinel):
+    with os.fdopen(fd, 'rb', closefd=False) as from_parent:
```

Comment:
Not very nice if `fd` and `parent_sentinel` are different fds... I think the caller should instead ensure those are different fds, possibly using `os.dup`.

Reply:
Woops, agreed.